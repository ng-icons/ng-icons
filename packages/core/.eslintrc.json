{
  "extends": ["../../.eslintrc.json"],
  "ignorePatterns": ["!**/*"],
  "overrides": [
    {
      "files": ["*.ts"],
      "extends": [
        "plugin:@nx/angular",
        "plugin:@angular-eslint/template/process-inline-templates"
      ],
      "rules": {
        "@angular-eslint/no-host-metadata-property": "off",
        "@typescript-eslint/no-inferrable-types": "off",
        "no-prototype-builtins": "off",
        "@typescript-eslint/no-non-null-assertion": "off",
        "@angular-eslint/directive-selector": [
          "error",
          {
            "type": "attribute",
            "prefix": "ng",
            "style": "camelCase"
          }
        ],
        "@angular-eslint/component-selector": [
          "error",
          {
            "type": "element",
            "prefix": "ng",
            "style": "kebab-case"
          }
        ]
      }
    },
    {
      "files": ["*.html"],
      "extends": ["plugin:@nx/angular-template"],
      "rules": {}
    }
  ]
}
