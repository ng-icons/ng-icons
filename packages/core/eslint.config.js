const { FlatCompat } = require('@eslint/eslintrc');
const baseConfig = require('../../eslint.config.js');
const js = require('@eslint/js');

const compat = new FlatCompat({
  baseDirectory: __dirname,
  recommendedConfig: js.configs.recommended,
});

module.exports = [
  ...baseConfig,
  ...compat
    .config({
      extends: [
        'plugin:@nx/angular',
        'plugin:@angular-eslint/template/process-inline-templates',
      ],
    })
    .map(config => ({
      ...config,
      files: ['**/*.ts'],
      rules: {
        '@angular-eslint/no-host-metadata-property': 'off',
        '@typescript-eslint/no-inferrable-types': 'off',
        'no-prototype-builtins': 'off',
        '@typescript-eslint/no-non-null-assertion': 'off',
        '@angular-eslint/directive-selector': [
          'error',
          {
            type: 'attribute',
            prefix: 'ng',
            style: 'camelCase',
          },
        ],
        '@angular-eslint/component-selector': [
          'error',
          {
            type: 'element',
            prefix: 'ng',
            style: 'kebab-case',
          },
        ],
        '@angular-eslint/component-class-suffix': 'off',
      },
    })),
  ...compat
    .config({ extends: ['plugin:@nx/angular-template'] })
    .map(config => ({
      ...config,
      files: ['**/*.html'],
      rules: {},
    })),
  ...compat.config({ parser: 'jsonc-eslint-parser' }).map(config => ({
    ...config,
    files: ['**/*.json'],
    rules: {
      '@nx/dependency-checks': [
        'error',
        {
          ignoredDependencies: [
            'tslib',
            '@eslint/eslintrc',
            '@eslint/js',
            '@ng-icons/feather-icons',
            'jest-preset-angular',
            '@angular/platform-browser',
            '@angular/router',
          ],
        },
      ],
    },
  })),
];
