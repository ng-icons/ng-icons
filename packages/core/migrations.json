{
  "generators": {
    "rename-icon-imports": {
      "version": "21.0.0",
      "description": "Rename all icons to lower camel case",
      "cli": "nx",
      "implementation": "./src/migrations/rename-icon-imports/rename-icon-imports"
    }
  },
  "schematics": {
    "rename-icon-imports": {
      "version": "21.0.0",
      "description": "Rename all icons to lower camel case",
      "cli": "ng",
      "implementation": "./src/migrations/rename-icon-imports/rename-icon-imports#schematic"
    }
  },
  "packageJsonUpdates": {
    "21.0.0": {
      "version": "21.0.0",
      "packages": {
        "@ng-icons/heroicons": {
          "version": "21.0.0",
          "alwaysAddToPackageJson": false
        },
        "@ng-icons/bootstrap-icons": {
          "version": "21.0.0",
          "alwaysAddToPackageJson": false
        },
        "@ng-icons/ionicons": {
          "version": "21.0.0",
          "alwaysAddToPackageJson": false
        },
        "@ng-icons/material-icons": {
          "version": "21.0.0",
          "alwaysAddToPackageJson": false
        },
        "@ng-icons/css.gg": {
          "version": "21.0.0",
          "alwaysAddToPackageJson": false
        },
        "@ng-icons/feather-icons": {
          "version": "21.0.0",
          "alwaysAddToPackageJson": false
        },
        "@ng-icons/jam-icons": {
          "version": "21.0.0",
          "alwaysAddToPackageJson": false
        },
        "@ng-icons/octicons": {
          "version": "21.0.0",
          "alwaysAddToPackageJson": false
        },
        "@ng-icons/radix-icons": {
          "version": "21.0.0",
          "alwaysAddToPackageJson": false
        },
        "@ng-icons/tabler-icons": {
          "version": "21.0.0",
          "alwaysAddToPackageJson": false
        },
        "@ng-icons/akar-icons": {
          "version": "21.0.0",
          "alwaysAddToPackageJson": false
        },
        "@ng-icons/iconoir": {
          "version": "21.0.0",
          "alwaysAddToPackageJson": false
        },
        "@ng-icons/cryptocurrency-icons": {
          "version": "21.0.0",
          "alwaysAddToPackageJson": false
        },
        "@ng-icons/simple-icons": {
          "version": "21.0.0",
          "alwaysAddToPackageJson": false
        },
        "@ng-icons/typicons": {
          "version": "21.0.0",
          "alwaysAddToPackageJson": false
        }
      }
    }
  }
}
