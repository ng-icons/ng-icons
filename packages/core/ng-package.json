{
  "$schema": "../../node_modules/ng-packagr/ng-package.schema.json",
  "dest": "../../dist/packages/core",
  "lib": {
    "entryFile": "src/index.ts"
  },
  "allowedNonPeerDependencies": ["@nrwl/devkit"]
}
