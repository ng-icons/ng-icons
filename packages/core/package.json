{
  "name": "@ng-icons/core",
  "version": "29.8.0",
  "license": "MIT",
  "keywords": [
    "angular",
    "icons",
    "ng-icons"
  ],
  "repository": {
    "url": "https://github.com/ng-icons/ng-icons"
  },
  "homepage": "https://ng-icons.github.io/ng-icons/",
  "peerDependencies": {
    "@angular/core": ">=18.0.0",
    "@angular/common": ">=18.0.0",
    "rxjs": "^6.5.3 || ^7.4.0"
  }
}
