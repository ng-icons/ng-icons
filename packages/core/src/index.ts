export * from './lib/components/glyph/glyph.component';
export * from './lib/components/icon/icon-name';
export * from './lib/components/icon/icon.component';
export * from './lib/icon.module';
export * from './lib/providers/glyph-config.provider';
export {
  NgGlyphs,
  NgGlyphset,
  provideNgGlyphs,
} from './lib/providers/glyph.provider';
export * from './lib/providers/icon-config.provider';
export * from './lib/providers/icon-loader.provider';
export * from './lib/providers/icon.provider';

// re-export the component as NgIconComponent to prevent breaking changes
export { NgIcon as NgIconComponent } from './lib/components/icon/icon.component';
