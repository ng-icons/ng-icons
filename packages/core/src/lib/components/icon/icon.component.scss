:host {
  display: inline-block;
  width: var(--ng-icon__size);
  height: var(--ng-icon__size);
  overflow: hidden;
}
