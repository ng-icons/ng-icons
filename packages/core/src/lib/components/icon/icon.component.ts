import {
  ChangeDetectionStrategy,
  Component,
  ElementRef,
  HostBinding,
  inject,
  Injector,
  Input,
  runInInjectionContext,
} from '@angular/core';
import type { IconName } from '../../components/icon/icon-name';
import { injectNgIconConfig } from '../../providers/icon-config.provider';
import {
  injectNgIconLoader,
  injectNgIconLoaderCache,
} from '../../providers/icon-loader.provider';
import { injectNgIcons } from '../../providers/icon.provider';
import { coerceLoaderResult } from '../../utils/async';
import { coerceCssPixelValue } from '../../utils/coercion';
import { toPropertyName } from '../../utils/format';

// This is a typescript type to prevent inference from collapsing the union type to a string to improve type safety
// eslint-disable-next-line @typescript-eslint/ban-types
export type IconType = IconName | (string & {});

@Component({
  selector: 'ng-icon',
  template: '',
  standalone: true,
  styleUrls: ['./icon.component.scss'],
  changeDetection: ChangeDetectionStrategy.OnPush,
})
export class NgIcon {
  /** Access the global icon config */
  private readonly config = injectNgIconConfig();

  /** Access the icons */
  private readonly icons = injectNgIcons();

  /** Access the icon loader if defined */
  private readonly loader = injectNgIconLoader();

  /** Access the icon cache if defined */
  private readonly cache = injectNgIconLoaderCache();

  /** Access the injector */
  private readonly injector = inject(Injector);

  /** Access the element ref */
  private readonly elementRef = inject<ElementRef<HTMLElement>>(ElementRef);

  /** Define the name of the icon to display */
  @Input() set name(name: IconType) {
    this.setIcon(name);
  }

  /** Define the size of the icon */
  @HostBinding('style.--ng-icon__size')
  @Input({ transform: coerceCssPixelValue })
  size?: string | number = this.config.size;

  /** Define the stroke-width of the icon */
  @HostBinding('style.--ng-icon__stroke-width')
  @Input()
  strokeWidth?: string | number;

  /** Define the color of the icon */
  @HostBinding('style.color')
  @Input()
  color?: string = this.config.color;

  /**
   * Load the icon with the given name and insert it into the template.
   * @param name The name of the icon to load.
   */
  private async setIcon(name: IconType): Promise<void> {
    const propertyName = toPropertyName(name);

    for (const icons of [...this.icons].reverse()) {
      if (icons[propertyName]) {
        // insert the SVG into the template
        this.elementRef.nativeElement.innerHTML = icons[propertyName];
        return;
      }
    }

    // if there is a loader defined, use it to load the icon
    if (this.loader) {
      const result = await this.requestIconFromLoader(name);

      // if the result is a string, insert the SVG into the template
      if (result !== null) {
        this.elementRef.nativeElement.innerHTML = result;
        return;
      }
    }

    // if there is no icon with this name warn the user as they probably forgot to import it
    console.warn(
      `No icon named ${name} was found. You may need to import it using the withIcons function.`,
    );
  }

  /**
   * Request the icon from the loader.
   * @param name The name of the icon to load.
   * @returns The SVG content for a given icon name.
   */
  private requestIconFromLoader(name: string): Promise<string> {
    return new Promise(resolve => {
      runInInjectionContext(this.injector, async () => {
        // if we have a cache, check if the icon is already loaded (i.e, it is a string)
        if (this.cache) {
          const cachedResult = this.cache.get(name);

          if (typeof cachedResult === 'string') {
            resolve(cachedResult);
            return;
          }

          // it may be a promise, so we need to await it
          if (cachedResult instanceof Promise) {
            const result = await cachedResult;
            resolve(result);
            return;
          }
        }

        const promise = coerceLoaderResult(this.loader!(name));

        // store the promise in the cache so if we get repeated calls (e.g. in a loop) before the loader has resolved
        // then don't call the loader function multiple times
        this.cache?.set(name, promise);

        // await the result of the promise
        const result = await promise;

        // if we have a cache, store the result
        this.cache?.set(name, result);

        resolve(result);
      });
    });
  }
}
