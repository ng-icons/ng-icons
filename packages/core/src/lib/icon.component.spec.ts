import { ComponentFixture, TestBed } from '@angular/core/testing';
import {
  FeatherAlertCircle,
  FeatherAlertTriangle,
} from '@ng-icons/feather-icons';
import { IconComponent } from './icon.component';
import { NgIconsModule } from './icon.module';

describe('Icon', () => {
  let component: IconComponent;
  let fixture: ComponentFixture<IconComponent>;
  let nativeElement: HTMLElement;

  beforeEach(async () => {
    await TestBed.configureTestingModule({
      imports: [
        NgIconsModule.withIcons({ FeatherAlertCircle, FeatherAlertTriangle }),
      ],
      declarations: [IconComponent],
    }).compileComponents();
  });

  beforeEach(() => {
    fixture = TestBed.createComponent(IconComponent);
    component = fixture.componentInstance;
    component.name = 'feather-alert-circle';
    fixture.detectChanges();
    nativeElement = fixture.nativeElement;
  });

  it('should insert the expected template', () => {
    expect(nativeElement).toMatchSnapshot();
  });

  it('should allow the icon to change', () => {
    component.name = 'feather-alert-triangle';
    fixture.detectChanges();
    expect(nativeElement).toMatchSnapshot();
  });
});
