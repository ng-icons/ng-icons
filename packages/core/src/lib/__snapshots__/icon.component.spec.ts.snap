// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Icon should allow the icon to change 1`] = `
<div
  id="root1"
  style="--ng-icon__size: 1em;"
>
  <svg
    class="feather feather-alert-triangle"
    fill="none"
    stroke="currentColor"
    stroke-linecap="round"
    stroke-linejoin="round"
    style="width:var(--ng-icon__size, 1em);height:var(--ng-icon__size, 1em);stroke-width:var(--ng-icon__stroke-width, 2)"
    viewBox="0 0 24 24"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      d="M10.29 3.86L1.82 18a2 2 0 0 0 1.71 3h16.94a2 2 0 0 0 1.71-3L13.71 3.86a2 2 0 0 0-3.42 0z"
    />
    <line
      x1="12"
      x2="12"
      y1="9"
      y2="13"
    />
    <line
      x1="12"
      x2="12.01"
      y1="17"
      y2="17"
    />
  </svg>
</div>
`;

exports[`Icon should insert the expected template 1`] = `
<div
  id="root0"
  style="--ng-icon__size: 1em;"
>
  <svg
    class="feather feather-alert-circle"
    fill="none"
    stroke="currentColor"
    stroke-linecap="round"
    stroke-linejoin="round"
    style="width:var(--ng-icon__size, 1em);height:var(--ng-icon__size, 1em);stroke-width:var(--ng-icon__stroke-width, 2)"
    viewBox="0 0 24 24"
    xmlns="http://www.w3.org/2000/svg"
  >
    <circle
      cx="12"
      cy="12"
      r="10"
    />
    <line
      x1="12"
      x2="12"
      y1="8"
      y2="12"
    />
    <line
      x1="12"
      x2="12.01"
      y1="16"
      y2="16"
    />
  </svg>
</div>
`;
